### ISSUE TITLE
Failing to invalidate download cache for `git+https` flake inputs

### ISSUE THREAD

**Issue by erictapen** [NixOS/nix#4743]
**Describe the bug**

For a while now I struggle with downloading new commits for `git+https` flake inputs if my flake was evaluated recently. That is especially annoying, as there seems to be no obvious way to override caching. All combinations of `--recreate-lock-file`, `--update-input`, or `--impure` won't help. So my usual solution is to just *wait* for the cache to be invalidated.

**Steps To Reproduce**

1. Have a flake.nix like this:

```nix
{
  inputs.sequoia.url =  "git+https://git.sr.ht/~erictapen/sequoia?ref=nix-flake";
  
  ...
}
```

2. Update the lock file, e.g. with `nix build --recreate-lock-file`

3. Add a commit to the repository at `git+https://git.sr.ht/~erictapen/sequoia`.

4. Rerun `nix build --recreate-lock-file` on the original flake, observe that the input is not updated.

**Expected behavior**

I'd expect Nix to never cache the old commit of an flake input, if `--recreate-lock-file` is specified.

**`nix-env --version` output**

`nix-env (Nix) 2.4pre20210423_fe2bf46` should be latest master.

**Additional context**

**Comment by Kha** [comment]
It's `--refresh`. I suggested `--recreate-lock-file` implying it in https://github.com/NixOS/nix/issues/3781#issuecomment-786670811, looks like people can definitely get confused otherwise.

**Comment by greedy** [comment]
> It's `--refresh`.

Yes, but `--refresh` has wider effects. I think that the download cache should be ignored for inputs being updated with `--update-input`.

**Comment by greedy** [comment]
If it still occurs, I still think the behavior described in this issue is confusing and annoying.
